### DISCUSSION TITLE
Zsh completion

### DISCUSSION THREAD

**Discussion by degnbol** [johnkerl/miller#1124] [Show and tell]
I made a quick draft for mlr completion for zsh at
https://github.com/degnbol/dotfiles/blob/main/miller/completion/_mlr

It is generated by parsing the outputs of several `mlr ... --help` and `mlr help` commands, see the code in the containig folder, e.g. RUNME.sh.

The main thing missing is support for `then` and `+` to chain multiple commands. If anyone knows more about zsh completion, let me know how to do that.

**Comment by aborruso** [comment]
Thank you

**Reply by degnbol** [reply]
Hope it's useful :)

**Comment by johnkerl** [comment]
@degnbol This is awesome!!!!! :)

A next step within Miller would be to have a command-line alternative like `mlr --generate-zsh-completions` which would create this kind of information automatically

**Reply by degnbol** [reply]
Have a look at the code at https://github.com/degnbol/completion_zsh_miller
I have used it for autogenerating :)

**Comment by J0J0** [comment]
@degnbol: Unfortunately, the above link is dead. Could you share your completion file (here)?

I understand, it might not be perfect or up-to-date; but it might be better than nothing! :)

**Reply by degnbol** [reply]
Here's a zip with the _mlr file for zsh and relevant code I used to generate it. The completion is not perfect, e.g. I haven't implemented completion for chained commands (with e.g. "+")
[completion.zip](https://github.com/user-attachments/files/20274504/completion.zip)

**Reply by eeweegh** [reply]
Hi @degnbol , 

Thanks for the _mlr file! Starting out with miller in zsh, and this comes in very handy.

just FYI, I tried to rebuild _mlr with the tools in that zip file, but it failed for two reasons:
1. the sed command in flags_help.sh is not compatible with the gnu sed version I had in brew, changing it into /usr/bin/sed made it work);
`sed: can't read /^--from/s/$/:filename:_files/: No such file or directory`
3. RUNME.sh makes a reference (using `git root`) to a SUB folder with python tools (like text_grep.py) - those are missing from the .zip file.

**Reply by degnbol** [reply]
Hi, thanks for checking it. For point 2, it should be fixed now, I edited the attached zip above. Specifically, changed it from a dependency to `text_grep.py` and instead just use `grep` and `sed`. Try to download it again and see.

**Reply by eeweegh** [reply]
hmm, there's still one text_grep.py references (as I dont know what it does exactly, cant fix it myself):
`./RUNME.zsh:    grep -v '^Usage:' $file | ./text_grep.py --stop "^Option|options:$|^$|^-" > $file.desc`

There's also three failing sed's (I'm using the gnu sed on Ubuntu and Darwin, I think that's because of the -i flag that behaves differently between those versions)
```
sed: can't read /^--from/s/$/:filename:_files/: No such file or directory
`sed: can't read /{.*file.*}/s/$/:filename:_files/: No such file or directory
`sed: can't read s/\[Show this message.\]//: No such file or directory
```

Then I also get this, but not sure where that comes from (I use mlr 6.13.0)
```
mlr: verb "mlr aux-list" not found. Please use "mlr --help" for a list.
mlr: verb "mlr hex" not found. Please use "mlr --help" for a list.
mlr: verb "mlr lecat" not found. Please use "mlr --help" for a list.
mlr: verb "mlr termcvt" not found. Please use "mlr --help" for a list.
mlr: verb "mlr unhex" not found. Please use "mlr --help" for a list.
```

**Reply by degnbol** [reply]
Hi again, I have made some improvements to this, try out the `_mlr` file now. I put everything for context at https://github.com/degnbol/completion_zsh_miller
It should now work with chaining (completion after "then" or "+") as well as completion of column names e.g. after `mlr --tsv --from FILE.tsv cut -f`.

**Comment by johnkerl** [comment]
Please also see #2096 which was included in the [Miller 6.20.2 release](https://github.com/johnkerl/miller/releases/tag/v6.20.2)
